File: flut/flutter/widgets/selection_container.py
from typing import Optional, override

from flut._flut.engine import (
    FlutAbstractObject,
    call_dart_static,
    named_constructor,
    _flut_pack_value,
)
from flut.flutter.widgets.framework import Widget


class SelectionRegistrar(FlutAbstractObject):
    _flut_type = "SelectionRegistrar"

    def add(self, selectable: "Selectable") -> None:
        raise NotImplementedError

    def remove(self, selectable: "Selectable") -> None:
        raise NotImplementedError

    @staticmethod
    def _flut_unpack(data: dict) -> "SelectionRegistrar":
        raise NotImplementedError(
            "SelectionRegistrar has no concrete wire form. Pass a concrete subtype."
        )

    @override
    def _flut_pack(self) -> dict[str, object]:
        raise NotImplementedError(
            "SelectionRegistrar has no concrete wire form. Pass a concrete subtype."
        )


class SelectionContainerDelegate(SelectionRegistrar):
    _flut_type = "SelectionContainerDelegate"

    @staticmethod
    def _flut_unpack(data: dict) -> "SelectionContainerDelegate":
        raise NotImplementedError(
            "SelectionContainerDelegate has no concrete wire form. Pass a concrete subtype."
        )

    @override
    def _flut_pack(self) -> dict[str, object]:
        raise NotImplementedError(
            "SelectionContainerDelegate has no concrete wire form. Pass a concrete subtype."
        )


class MultiSelectableSelectionContainerDelegate(SelectionContainerDelegate):
    _flut_type = "MultiSelectableSelectionContainerDelegate"

    @staticmethod
    def _flut_unpack(data: dict) -> "MultiSelectableSelectionContainerDelegate":
        raise NotImplementedError(
            "MultiSelectableSelectionContainerDelegate has no concrete wire form. Pass a concrete subtype."
        )

    @override
    def _flut_pack(self) -> dict[str, object]:
        raise NotImplementedError(
            "MultiSelectableSelectionContainerDelegate has no concrete wire form. Pass a concrete subtype."
        )


class StaticSelectionContainerDelegate(MultiSelectableSelectionContainerDelegate):
    _flut_type = "StaticSelectionContainerDelegate"

    def __init__(self) -> None:
        super().__init__()

    @staticmethod
    def _flut_unpack(data: dict) -> "StaticSelectionContainerDelegate":
        return StaticSelectionContainerDelegate()

    @override
    def _flut_pack(self) -> dict[str, object]:
        return self._flut_base_props()


class SelectionContainer(Widget):
    _flut_type = "SelectionContainer"

    def __init__(
        self,
        *,
        delegate: "SelectionContainerDelegate",
        child: Widget,
        key: Optional["Key"] = None,
        registrar: Optional["SelectionRegistrar"] = None,
    ) -> None:
        super().__init__(key=key)
        self.registrar = registrar
        self.delegate = delegate
        self.child = child

    @named_constructor
    def disabled(
        cls, *, child: Widget, key: Optional["Key"] = None
    ) -> "SelectionContainer":
        instance = cls.__new__(cls)
        Widget.__init__(instance, key=key)
        instance.registrar = None
        instance.delegate = None
        instance.child = child
        return instance

    @staticmethod
    def maybeOf(context: "BuildContext") -> Optional[SelectionRegistrar]:
        return call_dart_static("SelectionContainer", "maybeOf", context._flut_pack())

    @override
    def _flut_pack(self) -> dict[str, object]:
        result = self._flut_base_props()
        if self._flut_init != "disabled":
            if self.registrar is not None:
                result["registrar"] = _flut_pack_value(self.registrar)
            result["delegate"] = _flut_pack_value(self.delegate)
        result["child"] = _flut_pack_value(self.child)
        return result


class SelectionRegistrarScope(Widget):
    _flut_type = "SelectionRegistrarScope"

    def __init__(
        self,
        *,
        registrar: SelectionRegistrar,
        child: Widget,
        key: Optional["Key"] = None,
    ) -> None:
        super().__init__(key=key)
        self.registrar = registrar
        self.child = child

    @override
    def _flut_pack(self) -> dict[str, object]:
        result = self._flut_base_props()
        result["registrar"] = _flut_pack_value(self.registrar)
        result["child"] = _flut_pack_value(self.child)
        return result

File: flut/.flutter/lib/flutter/widgets/selection_container.dart
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/runtime.dart';

class FlutSelectionRegistrar extends FlutValueObject {
  const FlutSelectionRegistrar() : super('SelectionRegistrar');

  @override
  Map<String, dynamic> flutEncode() {
    throw UnimplementedError(
      'SelectionRegistrar has no concrete wire form. Pass a concrete subtype.',
    );
  }

  static SelectionRegistrar? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    throw UnimplementedError(
      'SelectionRegistrar has no concrete wire form. Pass a concrete subtype.',
    );
  }
}

class FlutSelectionContainerDelegate extends FlutValueObject {
  const FlutSelectionContainerDelegate() : super('SelectionContainerDelegate');

  @override
  Map<String, dynamic> flutEncode() {
    throw UnimplementedError(
      'SelectionContainerDelegate has no concrete wire form. Pass a concrete subtype.',
    );
  }

  static SelectionContainerDelegate? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    throw UnimplementedError(
      'SelectionContainerDelegate has no concrete wire form. Pass a concrete subtype.',
    );
  }
}

class FlutMultiSelectableSelectionContainerDelegate extends FlutValueObject {
  const FlutMultiSelectableSelectionContainerDelegate()
    : super('MultiSelectableSelectionContainerDelegate');

  @override
  Map<String, dynamic> flutEncode() {
    throw UnimplementedError(
      'MultiSelectableSelectionContainerDelegate has no concrete wire form. Pass a concrete subtype.',
    );
  }

  static MultiSelectableSelectionContainerDelegate? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    throw UnimplementedError(
      'MultiSelectableSelectionContainerDelegate has no concrete wire form. Pass a concrete subtype.',
    );
  }
}

class FlutStaticSelectionContainerDelegate extends FlutValueObject {
  final StaticSelectionContainerDelegate delegate;

  const FlutStaticSelectionContainerDelegate(this.delegate)
    : super('StaticSelectionContainerDelegate');

  @override
  Map<String, dynamic> flutEncode() => flutBaseProps();

  static StaticSelectionContainerDelegate? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return StaticSelectionContainerDelegate();
  }
}

class FlutSelectionContainer {
  FlutSelectionContainer._();

  static SelectionContainer? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    final constructor = runtime.unpackConstructor(data);
    switch (constructor) {
      case 'disabled':
        return SelectionContainer.disabled(
          key: runtime.decodeKey(data),
          child: runtime.unpackRequiredField<Widget>(data, 'child'),
        );
      default:
        return SelectionContainer(
          key: runtime.decodeKey(data),
          registrar: runtime.unpackOptionalField<SelectionRegistrar>(
            data,
            'registrar',
          ),
          delegate: runtime.unpackRequiredField<SelectionContainerDelegate>(
            data,
            'delegate',
          ),
          child: runtime.unpackRequiredField<Widget>(data, 'child'),
        );
    }
  }

  static void registerStatics(FlutRuntime runtime) {
    runtime.registerStatic('SelectionContainer.maybeOf', maybeOf);
  }

  static SelectionRegistrar? maybeOf(
    FlutRuntime runtime,
    BuildContext context,
  ) {
    return SelectionContainer.maybeOf(context);
  }
}

class FlutSelectionRegistrarScope {
  FlutSelectionRegistrarScope._();

  static SelectionRegistrarScope? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return SelectionRegistrarScope(
      key: runtime.decodeKey(data),
      registrar: runtime.unpackRequiredField<SelectionRegistrar>(
        data,
        'registrar',
      ),
      child: runtime.unpackRequiredField<Widget>(data, 'child'),
    );
  }
}
